### PR TITLE
Editor: Reduxify the EditorVisibility component 2nd attempt

### LINKED ISSUE #12276: Editor: Not able to edit passwords on password protected posts
It's not possible to edit passwords on password protected posts.

#### Steps to reproduce

1. Starting at: https://wordpress.com/post/
2. Publish a new post and, under the editor's Status module, edit the post's visibility so that it's password protected.
3. Navigate back to the editor for your newly published post.
4. Attempt to backspace over/edit the post's password.

#### What I expected

I expected to be able to edit the password after clicking/typing into the password field.

#### What happened instead

It wasn't possible to edit the password after typing straight into the field. 

I needed to toggle to "Public" and then back to "Password Protected" in order to be able to edit/change the password:

#### Browser / OS version

Chrome / macOS Sierra 10.12.3

#### Screenshot / Video

password

#### Context / Source

Reported in 249921-hc

#user-report #reallife

### PR DESCRIPTION
**Pull request by youknowriad**
This is the second attempt to close #12276

The first PR #12320 has been reverted because of the following use-case:

- Create a new post
- Fill quickly the title, content and set a password
- Hit publish before the first auto-save 
- The post remains as a draft and is not published

To fix this I switched the "status" priority on the `onPublish` method. The logic now is clearer.

- When publishing the status is set to "publish" unless the current edits contains a "publishable" status (private or future).

**Testing instructions**

- Try to publish privately
- Try to publish a future post
- Try to set/edit a password
- Try to revert to draft and republish

### CODE REVIEW

**Comment by alisterscott** [comment]
Thanks for this, I couldn't reproduce the last issue I reported where the post would not publish 👍 

There is still a race-condition I observed where if you fill in the details including the password and hit publish you will see the updated notice, then the published notice. Not as big of a deal as last time since it does publish, but I was able to reproduce seeing the updated notice before the publish notice a few times:

password 4

**Comment by jasmussen** [comment]
Nice work. I can't repeat the use case described, so that seems to have been fixed. 

Tested in the live branch in Chrome:

> Try to publish privately

Worked fine when changing a password protected post to private.

Worked fine when writing from scratch and publishing as private.

Worked fine when opening draft, publishing as private.

> Try to publish a future post

Wrote from scratch, scheduled a post for 2 minutes later. Published fine, even got a notification.

Tried publishing a post, then changing the date. Got a nice notification about the link changing:

screen shot 2017-03-21 at 10 55 20

Tried scheduling an existing and published post. Worked fine. 

> Try to set/edit a password
> Try to revert to draft and republish

Wrote from scratch a pw protected post. Worked fine. 

Tried editing a draft, adding a pw. Worked fine.

Tried reverting it back to draft. Worked fine, didn't clear password.

Tried republishing, worked fine.

Note — reverting to draft seemed to work a little bit unexpectedly for me. Here are the steps I took:

1. Write post from scratch
2. Set password
3. Publish
4. View post in one tab, blog index in other. Post showed up as PW protected in both
5. Revert to draft. Post disappeared from blog index, but post still showed up on the permalink URL in the other. Though was gone from a incognito mode.

In other words, reverting a pw protected post to draft feels like it put it in a state that's close to "Private". 

Post in question: https://testjoen.wordpress.com/2017/03/21/new-post/ — this shows up for me logged in, but not in incognito mode. Post in editor: https://calypso.live/post/testjoen.wordpress.com/446. Password is `pw`.

The behavior was the same when reverting to draft from a post without a password. This post should be a draft: https://testjoen.wordpress.com/2017/03/21/other-new-post/ — but I can still see it. Perhaps this behavior is intended?

**Comment by youknowriad** [comment]
> In other words, reverting a pw protected post to draft feels like it put it in a state that's close to "Private".

I think this behaviour is intended for all drafts. The logic is, since you can edit the post, you can preview it.

**Comment by jasmussen** [comment]
Everything seems fine to me!

**Comment by youknowriad** [comment]
@aduth do you have some time to look at this again?

**Comment by aduth** [comment]
I'm confused why, with #12320, the post would have been saved as a draft. Wouldn't the expected behavior be that upon selecting "Password protected", `edits.status` is assigned to "private" and takes precedence when publishing?

Or is the autosave kicking in and overriding it back to draft?

**Comment by aduth** [comment]
I think there might be bigger problems with how save functions in `PostActions` are considering the post status, but if we're moving these to be handled in Redux state, they're no longer accurate from the legacy `PostEditStore`.

**Comment by youknowriad** [comment]
> I'm confused why, with #12320, the post would have been saved as a draft

When we select "password protected", the updateVisibility callback is triggered which defaults to `draft` if the post is not saved (or autosaved) yet. See the logic here:

https://github.com/Automattic/wp-calypso/pull/12353/commits/f82e7ff1b766be9f69198a80befe210a559773dc#diff-104475887a82f7e64478bccf366fe981R170

So bringing the same "priority" back solves the issue. (Edits stats is overridden with 'publish' unless it's private or future)

**Comment by aduth** [comment]
I am very concerned about introducing anything to Redux state that operates on `status` unless we first remove all consideration from existing components and Flux action creators and stores.

**Comment by aduth** [comment]
Hmm, maybe it's not so much a problem in Flux actions, since we mirror edits from Redux to Flux before saving. This partial Reduxification is really quite confusing :confused:

**Comment by youknowriad** [comment]
> This partial Reduxification is really quite confusing

I understand, we'll have to decide if we want to continue trying this partial reduxification by steps or stop completely unless we reduxify the whole editor.

**Comment by youknowriad** [comment]
I did some more research regarding this ticket and what needs to be done in order to Reduxify the post status in the editor entirely, So I search for `status` and `postUtils` usage inside all the components in the post-editor. 

 - If it was used from the `savedPost`, I just ignored it, since this has no effect on the reduxification
 - I found some usage of these on `EditorPostDelete` and we can easily switch to the Redux Post Object here (instead of relying on the Flux post)
 - I found usage on `EditorSharingAccordion` but If we take a deeper look, the usage here should be based on the `savedPost` instead of the `post` with edits because we need to check if the post is effectively published (and not to be published), easy change as well
 - `EditorGroundControl` and `EditorPublishButton` makes a heavy usage of the status and they are currently based on the `Flux` post object. They also make heavy usage of the post's `date` which is really tied to the post status (check future). So I'm going to update the `date` edits and use Redux instead of `postActions` for this, that way the `status` and the `date` updates are on the same data, and I'll update these two components to rely on Redux instead of the Flux Post Object.

I think that's all what needs to be done to achieve this transition. Thoughts @aduth?

**Comment by aduth** [comment]
>So I'm going to update the date `edits` and use Redux instead of `postActions` for this

Are there then rippling effects on any components which rely on date?

Do you have a sense of whether the changes here alone would actually cause regressions? I had a hard time pin-pointing anything specifically that would break. Wondering if we could merge it in this admittedly fragile form so as to at least resolve the bug?

**Comment by youknowriad** [comment]
> Are there then rippling effects on any components which rely on date?

The effects here are small, just a small change on `PostEditor` and `EditorGroundControl`

> Do you have a sense of whether the changes here alone would actually cause regressions?

I'm having a hard time finding any regressions even. Maybe we should just merge it as for now to avoid big changes.
